docs/specs/add-kubernetes-support.md

```diff
@@ -0,0 +1,247 @@
+single path. It will call the cluster object asking for an operation. For
+example, to create a new Pod:
+```diff
+- k8scli.CoreV1().Pods(ns).Create(pod)
++ Cluster.CreatePod(name, image, extraVars, pullPolicy, serviceAccountName, ns)
+```
+
+The Cluster object will recieve the request and call the correct API.
+```diff
++ func CreatePod(name string, image spec.Image, extraVars string,
++                pullPolicy string, serviceAccountName string, ns string) {
++	pod := &v1.Pod{
++		ObjectMeta: metav1.ObjectMeta{
++			Name: apbID,
++		},
++		Spec: v1.PodSpec{
++			Containers: []v1.Container{
++				{
++					Name:  "apb",
++					Image: spec.Image,
++					Args: []string{
++						action,
++						"--extra-vars",
++						extraVars,
++					},
++					ImagePullPolicy: pullPolicy,
++				},
++			},
++			RestartPolicy:      v1.RestartPolicyNever,
++			ServiceAccountName: serviceAccountName,
++		},
++	}
++
++       k8scli.CoreV1().Pods(ns).Create(pod)
++
++ }
+
+```
+
+If there are identical API objects between COEs, check which cluster is
+being used.
+```diff
++ func CreateClusterRoleBinding(roleBindingName string, svcAccountName string,
++                               ApbRole string, ns string) {
++	roleBindingM := map[string]interface{}{
++		"apiVersion": "v1",
++		"kind":       "RoleBinding",
++		"metadata": map[string]string{
++			"name":      roleBindingName,
++			"namespace": ns,
++		},
++		"subjects": []map[string]string{
++			map[string]string{
++				"kind":      "ServiceAccount",
++				"name":      svcAccountName,
++				"namespace": ns,
++			},
++		},
++		"roleRef": map[string]string{
++			"name": ApbRole,
++		},
++      }
++
++      if OpenShift {
++         OCcli.CoreV1().ClusterRoleBinding(ns).Create(roleBindingM)
++      } elif Kubernetes {
++         k8scli.CoreV1().ClusterRoleBinding(ns).Create(roleBindingM)
++      }
++ }
+
+```
+
+### Hierarchy Diagram
+The diagram represents the code split into two files and the API objects
+in each of them
+
+**pkg/cluster/kubernetes.go**
+
+| API Resource | Uses API | Uses CLI | Different between OpenShift and Kubernetes |
+|:---:|:---:|:---:|:---:|
+| Pods | X |  |  |
+| Exec |  | X |  |
+| Login |  | X |  |
+| ServiceAccount |  | X |  |
+| RBAC ClusterRoleBindings |  | X | X |
+| RBAC ClusterRoles |  | X | X |
+| Namespace |  | X | X |
+
+**pkg/cluster/openshift.go**
+
+| API Resource | Uses API | Uses CLI | Different between OpenShift and Kubernetes |
+|:---:|:---:|:---:|:---:|
+| ClusterRoleBindings |  | X | X |
+| ClusterRoles |  | X | X |
+| Project |  | X | X |
+
+## APBs
+The APBs only target OpenShift as a cluster. The APBs will differ in two ways:
+ - Packages
+ - oc-login.sh
+
+### Packages
+The main issue with packaging is that folks using Kubernetes may only want the
+Kubernetes client. The package ```origin-clients``` installs both the OpenShift
+and Kubernetes clients.  If this the container size isn't an issue, then there
+may not be a problem using ```origin-clients```.
+
```

Comment:
How do we annotate that an APB requires OpenShift and can not run on Kubernetes.  Say it creates a Route or uses a DeploymentConfig?

I'm assuming any APB that runs on Kubernetes will be able to run on OpenShift.
Yet if an APB truly requires OpenShift (Route, DeploymentConfig, etc) it must have OpenShift and cannot run on pure k8s.

Would we want to allow an APB to downgrade itself based on environment?
i.e., it's written to deploy to OpenShift if available and use a DeploymentConfig, but if on K8S it follows a different path in the APB?